app/services/exchange_rates/upload_monthly_csv.rb

```diff
@@ -0,0 +1,47 @@
+module ExchangeRates
+  class UploadMonthlyCsv
+    delegate :instrument, to: ActiveSupport::Notifications
+
+    def self.call
+      new.call
+    end
+
+    def initialize
+      @current_time = Time.zone.now
+      @date_string = current_time.to_date.to_s
+      @month = current_time.month
+      @year = current_time.year
+    end
+
+    def call
+      # return unless penultimate_thursday?
+
+      csv_string = ExchangeRates::CreateCsv.call(data_result)
+
+      TariffSynchronizer::FileService.write_file(file_path, csv_string)
+
+      instrument('exchange_rates.monthly_csv',
+                 date: date_string,
+                 path: file_path,
+                 size: csv_string.size)
+    end
+
+  private
+
+    attr_reader :current_time, :month, :year, :date_string
+
+    def data_result
+      @data_result ||= ::ExchangeRateCurrencyRate.for_month(month, year)
+    end
+
+    def file_path
+      "data/exchange_rates/monthly_csv_#{date_string}.csv"
+    end
+
+    def penultimate_thursday?
```

Comment:
I love this. I wouldn't have approached it naturally like this but it makes loads of sense.

Reply:
Yeah, i was thinking could be done something like this but hadnt given much thought then realise @jebw had already solved it basically this way... So kudos to his idea!

Reply:
I'd probably have worked out every Thursday and then indexed the one we're targeting.